### PR TITLE
Refactoring Txp, part 2

### PR DESCRIPTION
**Pull request by gromakovsky**
Most notable changes:
1. `MapModifier` is used instead of `UtxoView`.
2. Txp payload and proof are now represented as `TxProof` and `TxPayload` types.
3. `_txInputs` and `_txOutputs` are now `NonEmpty` as well as `TxDistribution`.

It's not the end of CSL-826, but I want it to be merged now so that people use updated code.

### CODE REVIEW

File: src/Pos/Txp/MemState/Class.hs [hunk 1]
```diff
@@ -44,23 +40,38 @@ instance MonadTxpMem m => MonadTxpMem (StateT s m)
...
+        STM.readTVar txpUndos
+
+getMemPool :: (MonadIO m, MonadTxpMem m) => m MemPool
+getMemPool = getTxpLocalData (STM.readTVar . txpMemPool)
 
 modifyTxpLocalData :: (MonadIO m, MonadTxpMem m) => (TxpLocalDataPure -> (a, TxpLocalDataPure)) -> m a
 modifyTxpLocalData f =
     askTxpMem >>= \TxpLocalData{..} -> atomically $ do
-        curUV  <- STM.readTVar txpUtxoView
+        curUV  <- STM.readTVar txpUtxoModifier
```
Comment (pva701) [inline src/Pos/Txp/MemState/Class.hs]: I suppose `curUM`

File: src/Pos/Binary/Txp.hs [hunk 2]
```diff
@@ -53,3 +56,44 @@ instance Bi T.TxDistribution where
...
+        put _txpWitnesses
+        put _txpDistributions
+    get = do
+        _txpTxs <- get
+        _txpWitnesses <- get
+        _txpDistributions <- get
+        let lenTxs    = length _txpTxs
+            lenWit    = length _txpWitnesses
+            lenDistrs = length _txpDistributions
+        when (lenTxs /= lenWit) $ fail $ toString $
```
Comment (pva701) [inline src/Pos/Binary/Txp.hs]: Why do you check it here instead of creating constructor `mkTxPayload`?

Comment (gromakovsky) [inline src/Pos/Binary/Txp.hs]: Will be done in part 3 😉

File: src/Pos/Types/Arbitrary.hs [hunk 3]
```diff
@@ -280,10 +281,11 @@ derive makeArbitrary ''TxIn
 -- | Arbitrary transactions generated from this instance will only be valid
 -- with regards to 'mxTx'
 instance Arbitrary Tx where
-    arbitrary = do
-        txIns <- getNonEmpty <$> arbitrary
-        txOuts <- getNonEmpty <$> arbitrary
-        pure $ UnsafeTx txIns txOuts (mkAttributes ())
+    arbitrary =
+        mkTx <$> arbitrary <*> arbitrary <*>
+        pure (mkAttributes ()) <&> \case
```
Comment (gromakovsky) [inline src/Pos/Types/Arbitrary.hs]: `hindent` 🤷‍♂️

**Pull request merged**